Feature/20 appearing dialog

alipapapupu opened the pull request:
## Summary
<!--- Describe the change below, including rationale and design decisions -->
Created script that makes letters appear one at a time to the screen. You are able to change the speed of punctuation, dialog, or overall speed from the inky file. The code has function ready to speed up the dialog when player gives input, and this feature can be disabled/enabled from the inky file. You can prolong the dialog with next one after players input, or make the dialogs skip to next one without players input.

Does not change the speakers name or background color yet. For this I would request new PR implementing Scriptable Object containing the name, color etc. information about the speaker, and this could be used to changed all information on the screen.
 
## Before/after screenshots and/or animated gif
<!--- Skip this if not applicable -->
Example of the dialog appearing
![AppearingDialog](https://user-images.githubusercontent.com/16340242/131007291-c351d506-285a-48e2-98a7-23bb58e73603.gif)


## Testing instructions
<!--- What steps can be taken to manually verify the changes? -->
Open AppearingText-TestScene and hit play. The textbox contains Lorem Ipsum, but after you press space, the dialog starts. Take a look at the AppearingDialogInk-file while testing the dialog to see how they affect the way dialog appears.

## Additional information
<!--- Check any relevant boxes with "x" -->
- `[x]` Changes UI
- `[x]` Introduces new feature
- `[ ]` Removes existing feature
- `[x]` Has associated resource:  #20 
  <!--- Include any project card, github issue, etc. associated with this PR -->

alipapapupu committed:
Base functionality for the appearing text

The script has Coroutine you can call, that will continue until all  the dialog has been showed, and then return to caller.
The script also has public events, where you can add either in editor or via script events what should happen when dialog has reached its end.

alipapapupu committed:
Function to speed up on input

alipapapupu committed:
More complex text appear

Has own information for different situations like letter, dialog, overall, and punctuation.
Includes separating commands from dialog, but it was disabled for now

alipapapupu committed:
Made the appearing text work with Action Decoder

alipapapupu committed:
Fixed my unfinished part of DirectorActionDecoder

alipapapupu committed:
TextSkipping can be turned off

Players ability to skip or speed up text can be disabled if wanted, for example animation needs text in correct timing, so player speeding the text would be bad.

alipapapupu committed:
Made dialog controller talk with Appearing dialog controller

alipapapupu committed:
Dialogue controller now waits for appearing dialog to be ready

alipapapupu committed:
The dialog can be forced to appear right after previous one

alipapapupu committed:
Dialog can now be continued to the end of previous

alipapapupu committed:
Made disabling text skipping work with parameter

Previously disabling text skipping was 2 different commands. Now its one command with boolean parameter

alipapapupu committed:
Did needed steps for creating general textbox prefab

Added edited general textbox images
Made the code to edit them
Created prefab from them

alipapapupu committed:
Made event when letter appears

alipapapupu committed:
Example dialog to show appearing dialog in work

alipapapupu committed:
Letter determines the next letters appear time

Previously depending if the letter is punctuation or letter/digit determined how fast it itself is going to appear. This created effect that the punctuation itself is taking long time to appear. So now, the punctuation affects the next letters appear time, rather than its own, creating small pause before text continues after punctuation like it should be.

JuiceGrape commented on Assets/Resources/InkDialogueScripts/AppearingDialogInk.ink:
```diff
@@ -0,0 +1,46 @@
+BUT I DON'T WANNA WEAR PANTS!
+
+&OVERALL_SPEED:0.03
+&FORCE_NEXT_DIALOG:
```
I think it's better to replace this with "AUTO_SKIP:True" and "AUTO_SKIP:False". Just having auto skip enabled until we disable it seems like a more logical way of doing this than defining it for every line

JuiceGrape commented on Assets/Resources/InkDialogueScripts/AppearingDialogInk.ink:
```diff
@@ -0,0 +1,46 @@
+GOOD BECAUSE IF I WAS, I WOULD MAKE YOU GO TO BED WITHOUT ANY SHNITZEL!
+
+&CLEAR_SPEED:
+&DISABLE_SKIPPING:TRUE
```
What would DISABLE_SKIPPING:False do? For that matter, what does DISABLE_SKIPPING:True do?

alipapapupu replied:
DISABLE_SKIPPING:True makes it so player cant speed up or skip the dialog at all, rather, it will appear in constant speed. Good if dialog needs to be timed to appear with the animation, otherwise the player could speed up the dialog and screw things up. DISABLE_SKIPPING:False just makes the speeding possible again. If you have better names for this, I would be glad to hear them.

JuiceGrape replied:
Ah good thing to have indeed

JuiceGrape commented on Assets/Scripts/TextDecoder/DialogueController.cs:
```diff
@@ -67,12 +71,27 @@ public void OnNextLine()
+    ///<summary>
+    ///When the dialog has been written, makes the code be able to continue to next dialog
+    ///</summary>
+    public void StopWritingDialog()
```
Good way of solving this externally. However, button logic will not be handled in here, as it being in here was for debug purposes only (as denoted by the comment). Maybe create an "inputHandler" class which takes button presses and then triggers functions (through events) in the correct places? 
Basically I think it's best to have all input handling logic in it's own class, separate from the DialogueController

JuiceGrape replied:
Quick thought: We should probably make it a state machine

PhoebeMitchell replied:
UnityInputSystemInterface is now in the develop branch if you need an input class.

alipapapupu replied:
Fixed the issue. DialogueAppearController takes care of the autoSkip feature, and calls the OnNextLine when autoskip should happen. DialogueController also has bool that is changed by DialogueAppearControllers events to make sure dialog doesn't overlap, and it's not connected to the input in any way anymore.

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -13,6 +13,10 @@ public class DirectorActionDecoder : MonoBehaviour
     private IAudioController _audioController;
     private IEvidenceController _evidenceController;
 
+    [Header("Basic Values")]
```
Please remove these and add them like we have before. This class needs to be completely agnostic from anything external

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -143,7 +155,7 @@ void FadeInScene(string seconds)
         float timeInSeconds;
 
-        if(float.TryParse(seconds, out timeInSeconds))
+        if (float.TryParse(seconds, out timeInSeconds))
```
Please refrain from changes like this as they fog up git. Revert if you can

JuiceGrape replied:
In other places as well, I trust your judgement on finding them

alipapapupu replied:
Sure, my bad. I tend to auto format everything, and its in my muscle memory, but I'll see what I can do.

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -469,4 +481,84 @@ private bool HasEvidenceController()
+    /// Checks if the decoder has an appearing dialog controller attached, and shows an error if it doesn't
+    /// </summary>
+    /// <returns>Whether an appearing dialog controller is connected</returns>
+    private bool HasAppearingDialogController()
```
Move to controller stuff

JuiceGrape commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+using UnityEngine.Events;
+using UnityEngine.UI;
+
+public class AppearingDialogController : MonoBehaviour
```
Add interface

JuiceGrape commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+    ///<summary>
+    ///Events that should happen after completing a dialog. You have to manually add and remove listeners, and they will happen every time dialog is complete.
+    ///</summary>
+    public UnityEvent DialogDoneEvent { get { return _dialogDoneEvent; } }
```
Why does this exist? We should not be linking events in code, unless in special cases. This has no reason to be public

alipapapupu replied:
made private, and deleted extra event

JuiceGrape commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+    /// <param name="previousTime">How fast the dialog was going previously, in case the number should increase or decrease the current speed.</param>
+    /// <param name="newAppearTime">Gives out the new float containing new appear time rule.</param>
+    /// <returns>Returns true if numeric value was retrieved, or false it there was an error or the command didn't contain numeric value.</returns>
+    static private bool GetFloatFromString(string dialogText, float previousTime, out float newAppearTime)
```
This description makes no sense to me. What does this function do? The name is kinda bad too, cause getting a float from a string is just float.parse(string)

alipapapupu replied:
Legacy code from my original version, I can just delete the whole function.

JuiceGrape replied:
Legacy code should be removed unless you specifically intent to refactor or re-use it later

alipapapupu replied:
Refactored the method to work smarter way, and deleted unnecessary legacy code.

JuiceGrape commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+    ///<summary>
+    ///Small class containing if the current WaitTimer is in use, and how long to wait if it is.
+    ///</summary>
+    private class WaitInformation
```
Probably better as a struct?

alipapapupu replied:
Since the inUse and waitTime-values of the class is changed constantly, this being a struct won't work because that would require recreating the vakkye every time these values need to change.

JuiceGrape commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+    public void ContinueDialog()
+    {
+        _continueDialog = true;
+    }
```
Rest of the class looks great, keep up the good work.

JuiceGrape reviewed (changes requested):
Needs some work. Particularly sticking to existing code structures and moving logic out of existing classes where it doesn't belong.

PhoebeMitchell commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -469,4 +481,84 @@ private bool HasEvidenceController()
+        if (!HasAppearingDialogController())
+            return;
+
+        bool value = bool.Parse(disabled);
```
Consider using TryParse for error checking, and to be consistent with the rest of the script.

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+{
+    [Header("Basic Values")]
+    [SerializeField, Tooltip("TextMeshPro-component all the dialog should appear in.")]
+    private TextMeshProUGUI _controlledText = null;
```
nit: Do not need to initialise fields if the value is the field's default value.

alipapapupu replied:
I know, but Unity used to give warnings in the editor whenever a SerializeField or public value wasn't initialised in the code even though it would get initialised in the editor. They did fix this like a year ago, but it's already in my muscle memory to initialise values with the fields default value, and I will probably keep doing it even though it's a bit pointless.

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+        yield return StartDialogCoroutine(dialog);
+
+        _temporaryDialogDoneEvent?.Invoke();
+        _temporaryDialogDoneEvent.RemoveAllListeners();
```
Add ? for null check, like on previous line.

JuiceGrape replied:
What is the alternate course for a ? null check? If it's just ignoring the line, I'd prefer to make those null checks explicit in it's own line which aborts any future functionality

PhoebeMitchell replied:
For regular C# events it checks if the event has any listeners and is the same as checking "event != null".
I found out today though it's not necessary for Unity Events. '?' checks if the UnityEvent object itself is null, which it probably isn't if it's serialised.

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+    {
+        foreach (UnityAction ua in dialogDoneActions)
+        {
+            _temporaryDialogDoneEvent.AddListener(ua);
```
Could you just loop through the dialogDoneActions and invoke each one? Instead of subscribing them all to an event.

alipapapupu replied:
Removed the whole function since it was quite unnecessary

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+            WaiterTypes newWaiter = GetCurrentWaiter(_currentDialog[_currentLetterNum - 2]);
+
+            if (newWaiter == WaiterTypes.letter)
+                _allWaiters[WaiterTypes.letter].inUse = false;
```
Add braces to if statement.

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+    }
+
+    ///<summary>
+    ///Read the commands from dialog and seperates them. Prototype, currently not in use.
```
spelling: separates

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+                {
+                    startCommand = i;
+                }
+                else if (!insideCommand)
```
Use else instead of else if

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+    }
+
+    ///<summary>
+    ///Read the commands from dialog and seperates them. Prototype, currently not in use.
```
By prototype do you mean these methods will be used in future, or that they are no longer used? Can they be removed?

alipapapupu replied:
It depends on what we decide and what the designer team wants. Right now the speeds can be changed only before the dialog happens. Things markes as prototype would give full control over how the dialog is shown, giving the possibility to change speed values in the middle of the dialog and thus giving possibility to make more it complex, speeding and slowing when wanted. If we deem this unnecessary, I can remove everything marked as prototype, but I would like to hear yours and @JuiceGrape opinion on it, do we/design/writer team need that much control?

PhoebeMitchell replied:
I think this feature would be cool, but ask the design team what they're expecting. No point putting the effort in if they don't want it.

alipapapupu replied:
I'll leave these in for now, since there might be need for them later, but what we have right now seemed to be enough for people.

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+            if (newWaiter == WaiterTypes.letter)
+                _allWaiters[WaiterTypes.letter].inUse = false;
+
+            _currentAppearTime = _allWaiters[newWaiter].waitTime;
```
Punctuation speed affects things it probably shouldn't such as apostrophes, and symbols such as $ and £.

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+        //Is this information used only with punctuations.
+        public bool useOnlyWithPunctuations = false;
+        //Is this information used with everything.
+        public bool useWithEverything = false;
```
These should be properties, and in PascalCase

JuiceGrape replied:
Also, this seems like state behaviour, so it should be an enum instead of 2 bools. If neither of these can be true at the same time that is

PhoebeMitchell commented on Assets/Scripts/UI/AppearingDialogController.cs:
```diff
@@ -0,0 +1,517 @@
+    ///</summary>
+    private class WaitInformation
+    {
+        public WaitInformation() { }
```
Remove duplicate constructor.

PhoebeMitchell commented on Assets/Scripts/UI/WaiterTypes.cs:
```diff
@@ -0,0 +1,16 @@
+    defaultPunctuation,
+    dialog,
+    overall,
+    defaultValue,
```
Consider changing WaiterTypes to WaiterType and using PascalCase for values to follow C# conventions.
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces

PhoebeMitchell reviewed (approved):
Some suggestions, but overall the code looks functional, and runs as expected. AppearingDialogController is very big though. Could it be split up? It should also be be Dialogue instead of Dialog, as it's referring to conversation and not computer dialog boxes.

Not sure if having a mask to colour the textbox is necessary, when you could just set the colour of the mask image itself. I feel it would be easier to position text if the name box and the speech box were separate sprites. Then you could just get the speech text to fill the whole size of the speech box, and it would allow for easier position of the name text too.

Also correct the spelling of general in GerenralSpeechPanel. Then make the changes requested by JuiceGrape and it should be good.

alipapapupu committed:
Improved the the scripts according to comments

Created interface for AppearingDialogController
Made the AppearingDialogController and other edited scripts back to work completely indepentetly.
Fixed the GeneralSpeechPanel ui scaling.
Created canvas prefab
Improved some minor spelling and naming issues

alipapapupu committed:
ReadOnly attribute

alipapapupu committed:
Renamed AppearingDialogue and fixed my scene

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -468,5 +477,112 @@ private bool HasEvidenceController()
+            return true;
+        }
+
+        Debug.LogError("Bool value wasn't found from command. Please fix.");
```
Nondescript error message doesn't help anyone. Put the error message in the decode function, which sorta eliminates the need for having this a separate function.

alipapapupu committed:
Fixed autobuilder error

The autobuilder was causing error from editor script even though its in the editor folder. To get rid of the script, I put the editor script between #if UNITY_EDITOR so its forced not compile on anything else except editor.

alipapapupu committed:
Fixed minor bugs

Fixed minor bugs, split the dialog box images
Deleted unnecessary method
Mark a value as readonly
